vernac/vernacinterp.ml

```diff
@@ -13,89 +13,49 @@ open Synterp
-
-let with_succeed ~st f =
-  let () = ignore (f ()) in
-  Vernacstate.Interp.invalidate_cache ();
-  Vernacstate.unfreeze_full_state st;
-  if not !Flags.quiet
-  then Feedback.msg_notice Pp.(str "The command has succeeded and its effects have been reverted.")
-
 let locate_if_not_already ?loc (e, info) =
   (e, Option.cata (Loc.add_loc info) info (real_error_loc ~cmdloc:loc ~eloc:(Loc.get_loc info)))
 
-let interp_control_entry ~loc (f : control_entry) ~st
-    (fn : st:Vernacstate.t -> Vernacstate.LemmaStack.t option * Declare.OblState.t NeList.t) =
-  match f with
-  | ControlFail { st = synterp_st } ->
-    with_fail ~loc ~st (fun () -> Vernacstate.Synterp.unfreeze synterp_st; fn ~st);
-    st.Vernacstate.interp.lemmas, st.Vernacstate.interp.program
-  | ControlSucceed { st = synterp_st } ->
-    with_succeed ~st (fun () -> Vernacstate.Synterp.unfreeze synterp_st; fn ~st);
-    st.Vernacstate.interp.lemmas, st.Vernacstate.interp.program
-  | ControlTimeout { remaining } ->
-    vernac_timeout ~timeout:remaining (fun () -> fn ~st) ()
-  | ControlTime { synterp_duration } ->
-    let result = System.measure_duration (fun () -> fn ~st) () in
-    let result = Result.map (fun (v,d) -> v, System.duration_add d synterp_duration) result in
-    Feedback.msg_notice @@ System.fmt_transaction_result result;
-    begin match result with
-    | Ok (v,_) -> v
-    | Error (exn, _) -> Exninfo.iraise exn
-    end
-  | ControlInstructions { synterp_instructions } ->
-    let result = System.count_instructions (fun () -> fn ~st) () in
-    let result = Result.map (fun (v,d) -> v, System.instruction_count_add d synterp_instructions) result in
-    Feedback.msg_notice @@ System.fmt_instructions_result result;
-    begin match result with
-    | Ok (v,_) -> v
-    | Error (exn, _) -> Exninfo.iraise exn
-    end
-  | ControlRedirect s ->
-    Topfmt.with_output_to_file s (fun () -> fn ~st) ()
-
-(* "locality" is the prefix "Local" attribute, while the "local" component
- * is the outdated/deprecated "Local" attribute of some vernacular commands
- * still parsed as the obsolete_locality grammar entry for retrocompatibility.
- * loc is the Loc.t of the vernacular command being interpreted. *)
-let rec interp_expr ?loc ~atts ~st c =
+let with_interp_state ~unfreeze_transient st =
+  let with_local_state synterp_st f =
+    unfreeze_transient synterp_st;
+    let v = f () in
+    Vernacstate.Interp.invalidate_cache ();
+    Vernacstate.unfreeze_full_state st;
+    (), v
+  in
+  { VernacControl.with_local_state }
+
+let interp_control_gen ~loc ~st ~unfreeze_transient control f =
+  let noop = st.Vernacstate.interp.lemmas, st.Vernacstate.interp.program in
+  let control, res =
+    VernacControl.under_control ~loc
+      ~with_local_state:(with_interp_state ~unfreeze_transient st)
+      control
+      ~noop
+      f
+  in
+  if VernacControl.after_last_phase ~loc control
+  then noop
```

Comment:
the handling of unchanged functionally passed state here feels a bit awkward so maybe the api could be improved a bit
Not sure how though